constraints.pro

```diff
@@ -0,0 +1,222 @@
+gen_enforced_field(WorkspaceCwd, 'homepage', null) :-
+  workspace_field(WorkspaceCwd, 'private', true).
+
+% "repository.type" must be "git" for all packages.
+gen_enforced_field(WorkspaceCwd, 'repository.type', 'git').
+
+% "repository.url" must be "https://github.com/MetaMask/controllers.git" for all
+% packages.
+gen_enforced_field(WorkspaceCwd, 'repository.url', 'https://github.com/MetaMask/controllers.git').
+
+% "license" must be "MIT" for all workspace packages and unset for the root.
+gen_enforced_field(WorkspaceCwd, 'license', 'MIT') :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
+gen_enforced_field(WorkspaceCwd, 'license', null) :-
+  workspace_field(WorkspaceCwd, 'private', true).
+
+% "main" must be "dist/index.js" for workspace packages and unset for the
+% root.
+gen_enforced_field(WorkspaceCwd, 'main', './dist/index.js') :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
+gen_enforced_field(WorkspaceCwd, 'main', null) :-
+  workspace_field(WorkspaceCwd, 'private', true).
+
+% "types" must be "dist/index.d.ts" for workspace packages and unset for the
+% root.
+gen_enforced_field(WorkspaceCwd, 'types', './dist/index.d.ts') :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
+gen_enforced_field(WorkspaceCwd, 'types', null) :-
+  workspace_field(WorkspaceCwd, 'private', true).
+
+% "files" must be ["dist/"] for workspace packages and unset for the root.
+gen_enforced_field(WorkspaceCwd, 'files', ['dist/']) :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
+gen_enforced_field(WorkspaceCwd, 'files', null) :-
+  workspace_field(WorkspaceCwd, 'private', true).
+
+% "engines.node" must be ">=14.0.0" for all packages.
+gen_enforced_field(WorkspaceCwd, 'engines.node', '>=14.0.0').
+
+% "publishConfig.access" must be "public" for workspace packages and unset
+% for the root.
+gen_enforced_field(WorkspaceCwd, 'publishConfig.access', 'public') :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
+gen_enforced_field(WorkspaceCwd, 'publishConfig.access', null) :-
+  workspace_field(WorkspaceCwd, 'private', true).
+
+% "publishConfig.registry" must be "https://registry.npmjs.org" for all
+% workspace packages and unset for the root.
+gen_enforced_field(WorkspaceCwd, 'publishConfig.registry', 'https://registry.npmjs.org/') :-
+  \+ workspace_field(WorkspaceCwd, 'private', true).
+gen_enforced_field(WorkspaceCwd, 'publishConfig.registry', null) :-
+  workspace_field(WorkspaceCwd, 'private', true).
+
+% All dependency ranges must be recognizable.
+gen_enforced_dependency(WorkspaceCwd, DependencyIdent, 'a range optionally starting with ^ or ~', DependencyType) :-
+  workspace_has_dependency(WorkspaceCwd, DependencyIdent, DependencyRange, DependencyType),
+  \+ is_valid_version_range(DependencyRange).
+
+% All dependency ranges for a package must be synchronized across the monorepo
+% (the least version range wins), regardless of which "*dependencies" the
```

Comment:
@Gudahtt Wanted to make sure you saw this — I just changed this from greatest wins to least wins as it ensures that if we pin a dependency to a specific past version, that pin doesn't get wiped away later accidentally.

Reply:
Makes sense!